programs/eosio-blocklog/main.cpp

```diff
@@ -169,6 +176,176 @@ void blocklog::initialize(const variables_map& options) {
+   if (version == 1)
+      firstBlock= 1;
+   else
+      fin0.sgetn((char*)&firstBlock,sizeof(firstBlock));
+   cout << "version= " << version << "\nfirst block= " << firstBlock << '\n';
+   if (n <= firstBlock) {
+      cout << n << " is before first block so nothing to do\n";
+      return 2;
+   }
+
+   //open blocks.index and get blocks.log position for block 'n'
+   filebuf fin1;
+   string indexFileName= (blockDir/"blocks.index").generic_string();
+   fin1.open(indexFileName.c_str(),ios::in|ios::binary);
+   EOS_ASSERT( fin1.is_open(), block_index_not_found, "cannot read blocks.index" );
+   uint64_t indexPos= 8*(n-firstBlock);
+   uint64_t pos= fin1.pubseekoff(indexPos,ios::beg,ios::in);
+   EOS_ASSERT( pos==indexPos, block_log_exception, "cannot read blocks.index entry for trim-after-block" );
+   uint64_t fpos0, fpos1;                                   //filepos of block n and block n+1, will read from blocks.index
+   fin1.sgetn((char*)&fpos0,sizeof(fpos0));
+   fin1.sgetn((char*)&fpos1,sizeof(fpos1));
+   fin1.close();
+   cout << "According to blocks.index:\n";
+   cout << "    block " << n << " starts at position " << fpos0 << '\n';
+   cout << "    block " << n+1 << " starts at position " << fpos1 << '\n';
+
+   //read blocks.log and verify block number n is found at file position fpos0
+   fin0.pubseekoff(fpos0,ios::beg,ios::in);
+   fin0.sgetn((char*)&blkStart,sizeof(blkStart));
+   fin0.close();
+   uint32_t bnum= endian_reverse_u32(blkStart.blknum)+1;    //convert from big endian to little endian, add 1 since prior block
+   EOS_ASSERT( bnum==n, block_log_exception, "blocks.index does not agree with blocks.log" );
+   cout << "In blocks.log at position " << fpos0 << " find block " << bnum << " as expected\n";
+   EOS_ASSERT( truncate(blockFileName.c_str(),fpos1)==0, block_log_exception, "truncate blocks.log fails");
+   indexPos+= sizeof(uint64_t);                             //advance to after record for block n
+   EOS_ASSERT( truncate(indexFileName.c_str(),indexPos)==0, block_log_exception, "truncate blocks.index fails");
+   cout << "blocks.log has been truncated to " << fpos1 << " bytes\n";
+   cout << "blocks.index has been truncated to " << indexPos << " bytes\n";
+   return 0;
+}
+
+int makeIndex(bfs::path blockDir, string outFile) {
+   //this code makes blocks.index much faster than nodeos (in recent test 80 seconds vs. 90 minutes)
+   using namespace std;
+   cout << "Will make blocks.index from blocks.log\n";
+   string blockFileName= (blockDir / "blocks.log").generic_string();
+   string outFileName=   (blockDir / outFile     ).generic_string();
+   int fin = open(blockFileName.c_str(), O_RDONLY);
+   EOS_ASSERT( fin>0, block_log_not_found, "cannot read blocks.log" );
+
+   //will read big chunks of blocks.log into buf, will fill fposList with file positions before write to blocks.index
+   constexpr uint32_t bufLen{1U<<24};                       //bufLen must be power of 2 >= largest possible block == one MByte
```

Comment:
This is not strictly true. A block can be bigger than 1 MByte. The max is configurable and even the default of 1 MByte is only a soft max. `libraries/chain/include/eosio/chain/chain_config.hpp` `max_block_net_usage`

Reply:
Just some notes for me.
1<<24 is 16,777,216
1<<22 is 4,194,304
fposListLen >> 3 is 524,288 which is 4,194,304